framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -949,146 +951,175 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
-                            internalReceiver, null, checker.getContext());
+                if (!checkContract(expr, anno, inferredAnno, store)) {
+                    checker.report(Result.failure(
+                            treeForErrorReporting.getKind() == Tree.Kind.METHOD_INVOCATION
+                                ? "contracts.precondition.not.satisfied"
+                                : "contracts.precondition.not.satisfied.field",
+                            treeForErrorReporting.toString(),
+                            expr == null ? expression : expr.toString()), treeForErrorReporting);
                 }
+            } catch (FlowExpressionParseException e) {
+                // errors are reported at declaration site
             }
+        }
+    }
 
-            if (flowExprContext != null) {
-                FlowExpressions.Receiver expr = null;
-                try {
-                    CFAbstractStore<?, ?> store = atypeFactory.getStoreBefore(tree);
+    private FlowExpressionContext getFlowExpressionContextFromNode(Node node) {
+        FlowExpressionContext flowExprContext = null;
 
-                    String s = expression.trim();
-                    Pattern selfPattern = Pattern.compile("^(this)$");
-                    Matcher selfMatcher = selfPattern.matcher(s);
-                    if (selfMatcher.matches()) {
-                        s = flowExprContext.receiver.toString(); // it is possible that s == "this" after this call
-                    }
+        if (node instanceof MethodInvocationNode) {
+            flowExprContext = FlowExpressionParseUtil
+                    .buildFlowExprContextForUse(
+                            (MethodInvocationNode) node, checker.getContext());
+        }
+        else if (node instanceof FieldAccessNode) {
+            // Adapted from FlowExpressionParseUtil.buildFlowExprContextForUse
 
-                    expr = FlowExpressionParseUtil.parse(expression,
-                            flowExprContext, getCurrentPath());
+            Receiver internalReceiver = FlowExpressions.internalReprOf(atypeFactory,
+                ((FieldAccessNode) node).getReceiver());
 
-                    if (expr == null) {
-                        // TODO: Wrap the following 'itself' handling logic into a method that calls FlowExpressionParseUtil.parse
+            flowExprContext = new FlowExpressionContext(
+                    internalReceiver, null, checker.getContext());
+        }
+        else if (node instanceof LocalVariableNode) {
+            // Adapted from org.checkerframework.dataflow.cfg.CFGBuilder.CFGTranslationPhaseOne.visitVariable
 
-                        /** Matches 'itself' - it refers to the variable that is annotated, which is different from 'this' */
-                        Pattern itselfPattern = Pattern.compile("^itself$");
-                        Matcher itselfMatcher = itselfPattern.matcher(expression.trim());
+            ClassTree enclosingClass = TreeUtils
```

Comment:
I would put the class name and the static method on the same line, here and in the next statement. Neither line is overly long.